cpp/CMakeLists.txt

```diff
@@ -382,6 +375,38 @@ enable_testing()
 ############################################################
 
 # ----------------------------------------------------------------------
+# Thirdparty toolchain
+
+set(THIRDPARTY_DIR "${CMAKE_SOURCE_DIR}/thirdparty")
+set(GFLAGS_VERSION "2.1.2")
+set(GTEST_VERSION "1.8.0")
+set(GBENCHMARK_VERSION "1.1.0")
+set(FLATBUFFERS_VERSION "1.6.0")
+set(JEMALLOC_VERSION "4.4.0")
+
+set(FLATBUFFERS_HOME "$ENV{FLATBUFFERS_HOME}")
+set(RAPIDJSON_HOME "$ENV{RAPIDJSON_HOME}")
+set(JEMALLOC_HOME "$ENV{JEMALLOC_HOME}")
+
+if (NOT "$ENV{ARROW_BUILD_TOOLCHAIN}" STREQUAL "")
```

Comment:
I would turn the logic around here. If `ARROW_BUILD_TOOLCHAIN` is set, use it. But if a more specific environment variable like `JEMALLOC_HOME` is set, it should override `ARROW_BUILD_TOOLCHAIN`. Thus you can use e.g. conda for all except `jemalloc` which you want to use the system provided one.